R/semicolon_linter.R

```diff
@@ -12,36 +12,44 @@
 #'   <https://style.tidyverse.org/syntax.html#semicolons>
 #' @export
 semicolon_linter <- function(allow_compound = FALSE, allow_trailing = FALSE) {
-  Linter(function(source_expression) {
-    tokens <- with_id(source_expression, ids_with_token(source_expression, "';'"))
-    is_trailing <- is_trailing_sc(tokens, source_expression)
+  msg_trailing <- "Trailing semicolons are not needed."
+  msg_compound <- "Compound semicolons are discouraged. Replace them by a newline."
 
-    to_keep <- (is_trailing & !allow_trailing) |
-               (!is_trailing & !allow_compound)
+  if (allow_compound && allow_trailing) {
+    stop("At least one of `allow_compound` or `allow_trailing` must be FALSE, otherwise no lints can be generated.")
+  } else if (allow_compound && !allow_trailing) {
+    # lint only trailing
+    xpath <- "//OP-SEMICOLON[not(@line1 = following-sibling::*[1]/@line1)]"
+    msg <- msg_trailing
```

Comment:
false positive in object usage linter?

Reply:
Yes, so it seems.
Instead of nolinting, we could unconditionally use the `ifelse()` without losing correctness. What do you prefer?

Reply:
Might be hard to fix on our end though because checkUsage() misreports.

Reply:
I see... something to report with `codetools` then?

https://gitlab.com/luke-tierney/codetools/-/issues

Reply:
yes, `object_usage_linter()` is just a pretty wrapper around `codetools::checkUsage()`.